### PR TITLE
linuxPackages.ena: 2.12.0 -> 2.12.3; general maintenance

### PR DESCRIPTION
**Pull request by kiike**
## Description of changes

This PR reworks the packaging of ena to improve some actions carried out on #310680.

- instead of patching the `Kcompat.h` file to resolve building, remove an unneeded flag that breaks kernel feature detection. This flag was introduced for 3.12, which is a kernel version that NixOS doesn't support anymore.
- replace the current patch, which targets fixing building for all supported versions, for one that only targets 4.19, the only version that will not build with the vanilla sources. This patch comes from upstream and is temporary until the next release.
- reformat using nixfmt-rfc-style

EDIT 5 June: upgrade to 2.12.1

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by kiike** [comment]
Result of `nixpkgs-review pr 313941` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>22 packages built:</summary>
  <ul>
    <li>linuxKernel.packages.linux_4_19.ena</li>
    <li>linuxKernel.packages.linux_4_19_hardened.ena</li>
    <li>linuxKernel.packages.linux_5_10.ena</li>
    <li>linuxKernel.packages.linux_5_10_hardened.ena</li>
    <li>linuxKernel.packages.linux_5_15.ena</li>
    <li>linuxKernel.packages.linux_5_15_hardened.ena</li>
    <li>linuxKernel.packages.linux_5_4.ena</li>
    <li>linuxKernel.packages.linux_5_4_hardened.ena</li>
    <li>linuxKernel.packages.linux_6_1.ena</li>
    <li>linuxKernel.packages.linux_6_1_hardened.ena</li>
    <li>linuxKernel.packages.linux_6_6.ena</li>
    <li>linuxKernel.packages.linux_hardened.ena (linuxKernel.packages.linux_6_6_hardened.ena)</li>
    <li>linuxKernel.packages.linux_6_8.ena</li>
    <li>linuxKernel.packages.linux_6_8_hardened.ena</li>
    <li>linuxKernel.packages.linux_6_9.ena</li>
    <li>linuxKernel.packages.linux_6_9_hardened.ena</li>
    <li>linuxKernel.packages.linux_latest_libre.ena</li>
    <li>linuxKernel.packages.linux_libre.ena</li>
    <li>linuxKernel.packages.linux_lqx.ena</li>
    <li>linuxKernel.packages.linux_xanmod.ena</li>
    <li>linuxKernel.packages.linux_xanmod_latest.ena (linuxKernel.packages.linux_xanmod_stable.ena)</li>
    <li>linuxKernel.packages.linux_zen.ena</li>
  </ul>
</details>

**Comment by kiike** [comment]
Converting to draft while I try upgrading the package.

**Comment by IldenH** [comment]
2.12.3 is out. Upgrade straight to that instead?
https://github.com/amzn/amzn-drivers/releases/tag/ena_linux_2.12.3.

**Review by sielicki — APPROVED** [review]

File: pkgs/os-specific/linux/ena/default.nix [hunk 1]
```diff
@@ -47,7 +50,10 @@ stdenv.mkDerivation rec {
     description = "Amazon Elastic Network Adapter (ENA) driver for Linux";
     homepage = "https://github.com/amzn/amzn-drivers";
     license = licenses.gpl2Only;
-    maintainers = with maintainers; [ eelco sielicki ];
+    maintainers = with maintainers; [
+      eelco
+      sielicki
```
Comment (Mic92) [inline pkgs/os-specific/linux/ena/default.nix:55]: Arianvp agreed:

```suggestion
      sielicki
      arianvp
```

**Review by arianvp — APPROVED** [review]

**Comment by arianvp** [comment]
Hmm:

Result of `nixpkgs-review pr 313941` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>5 packages marked as broken and skipped:</summary>
  <ul>
    <li>linuxKernel.packages.linux_lqx.ena</li>
    <li>linuxKernel.packages.linux_xanmod.ena</li>
    <li>linuxKernel.packages.linux_xanmod_latest.ena</li>
    <li>linuxKernel.packages.linux_xanmod_stable.ena</li>
    <li>linuxKernel.packages.linux_zen.ena</li>
  </ul>
</details>
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>linuxKernel.packages.linux_6_10.ena</li>
    <li>linuxKernel.packages.linux_latest_libre.ena</li>
  </ul>
</details>
<details>
  <summary>17 packages built:</summary>
  <ul>
    <li>linuxKernel.packages.linux_4_19.ena</li>
    <li>linuxKernel.packages.linux_4_19_hardened.ena</li>
    <li>linuxKernel.packages.linux_5_10.ena</li>
    <li>linuxKernel.packages.linux_5_10_hardened.ena</li>
    <li>linuxKernel.packages.linux_5_15.ena</li>
    <li>linuxKernel.packages.linux_5_15_hardened.ena</li>
    <li>linuxKernel.packages.linux_5_4.ena</li>
    <li>linuxKernel.packages.linux_5_4_hardened.ena</li>
    <li>linuxKernel.packages.linux_6_1.ena</li>
    <li>linuxKernel.packages.linux_6_1_hardened.ena</li>
    <li>linuxKernel.packages.linux_6_6.ena</li>
    <li>linuxKernel.packages.linux_hardened.ena (linuxKernel.packages.linux_6_6_hardened.ena)</li>
    <li>linuxKernel.packages.linux_6_8.ena</li>
    <li>linuxKernel.packages.linux_6_8_hardened.ena</li>
    <li>linuxKernel.packages.linux_6_9.ena</li>
    <li>linuxKernel.packages.linux_6_9_hardened.ena</li>
    <li>linuxKernel.packages.linux_libre.ena</li>
  </ul>
</details>

**Comment by arianvp** [comment]
The error in question is

```
make -C /nix/store/2yg16mg5cgcckxl0hjzvhz7sw3yrkwjm-linux-6.10.1-dev/lib/modules/6.10.1/build M=/build/source/kernel/linux/ena modules
make[1]: Entering directory '/nix/store/2yg16mg5cgcckxl0hjzvhz7sw3yrkwjm-linux-6.10.1-dev/lib/modules/6.10.1/build'
  CC [M]  /build/source/kernel/linux/ena/ena_netdev.o
  CC [M]  /build/source/kernel/linux/ena/ena_ethtool.o
  CC [M]  /build/source/kernel/linux/ena/ena_lpc.o
  CC [M]  /build/source/kernel/linux/ena/ena_phc.o
  CC [M]  /build/source/kernel/linux/ena/ena_xdp.o
/build/source/kernel/linux/ena/ena_xdp.c: In function 'ena_xdp_rx_skb_zc':
/build/source/kernel/linux/ena/ena_xdp.c:749:15: error: implicit declaration of function '__napi_alloc_skb'; did you mean 'napi_alloc_skb'? [-Werror=implicit-function-declaration]
  749 |         skb = __napi_alloc_skb(rx_ring->napi,
      |               ^~~~~~~~~~~~~~~~
      |               napi_alloc_skb
compilation terminated due to -Wfatal-errors.
cc1: some warnings being treated as errors
make[3]: *** [/nix/store/2yg16mg5cgcckxl0hjzvhz7sw3yrkwjm-linux-6.10.1-dev/lib/modules/6.10.1/source/scripts/Makefile.build:244: /build/source/kernel/linux/ena/ena_xdp.o] Error 1
make[2]: *** [/nix/store/2yg16mg5cgcckxl0hjzvhz7sw3yrkwjm-linux-6.10.1-dev/lib/modules/6.10.1/source/Makefile:1934: /build/source/kernel/linux/ena] Error 2
make[1]: *** [/nix/store/2yg16mg5cgcckxl0hjzvhz7sw3yrkwjm-linux-6.10.1-dev/lib/modules/6.10.1/source/Makefile:240: __sub-make] Error 2
make[1]: Leaving directory '/nix/store/2yg16mg5cgcckxl0hjzvhz7sw3yrkwjm-linux-6.10.1-dev/lib/modules/6.10.1/build'
make: *** [Makefile:102: ena.ko] Error 2
```

**Comment by Mic92** [comment]
I think just marking the packages as broken on 6.10 until ena has been updated, should be enough.

**Comment by arianvp** [comment]
I think this is ready to go.

The patch is by me from https://github.com/amzn/amzn-drivers/issues/313#issuecomment-2254386304

They'll provide a modified version if it in the next release but I prefer not waiting for that.

The XDP support is experimental anyways so most people aren't using this codepath. 

If no complaints I'm gonna merge this

**Comment by arianvp** [comment]
Okay fixing this patch surfaces more errors. How do we feel about marking this broken on 6.10?

```
error: builder for '/nix/store/lxmbs6ya60x68wf5a8mpk5njg3f0xrwh-ena-2.12.3-6.10.2.drv' failed with exit code 2;
       last 10 log lines:
       > /build/source/kernel/linux/ena/ena_xdp.c: In function 'ena_xdp_clean_rx_irq_zc':
       > /build/source/kernel/linux/ena/ena_xdp.c:816:17: error: too many arguments to function 'xsk_buff_dma_sync_for_cpu'
       >   816 |                 xsk_buff_dma_sync_for_cpu(xdp, rx_ring->xsk_pool);
       >       |                 ^~~~~~~~~~~~~~~~~~~~~~~~~
       > compilation terminated due to -Wfatal-errors.
       > make[3]: *** [/nix/store/216k8yzq05s5r3lvnh7jd9hbric4cmln-linux-libre-6.10.2-dev/lib/modules/6.10.2-gnu/source/scripts/Makefile.build:244: /build/source/kernel/linux/ena/ena_xdp.o] Error 1
       > make[2]: *** [/nix/store/216k8yzq05s5r3lvnh7jd9hbric4cmln-linux-libre-6.10.2-dev/lib/modules/6.10.2-gnu/source/Makefile:1934: /build/source/kernel/linux/ena] Error 2
       > make[1]: *** [/nix/store/216k8yzq05s5r3lvnh7jd9hbric4cmln-linux-libre-6.10.2-dev/lib/modules/6.10.2-gnu/source/Makefile:240: __sub-make] Error 2
       > make[1]: Leaving directory '/nix/store/216k8yzq05s5r3lvnh7jd9hbric4cmln-linux-libre-6.10.2-dev/lib/modules/6.10.2-gnu/build'
       > make: *** [Makefile:102: ena.ko] Error 2
       For full logs, run 'nix log /nix/store/lxmbs6ya60x68wf5a8mpk5njg3f0xrwh-ena-2.12.3-6.10.2.drv'.
error: builder for '/nix/store/j23sjz0wq8gq3y9p8a6082r0nvg9qnyi-ena-2.12.3-6.10.2.drv' failed with exit code 2;
       last 10 log lines:
       > /build/source/kernel/linux/ena/ena_xdp.c: In function 'ena_xdp_clean_rx_irq_zc':
       > /build/source/kernel/linux/ena/ena_xdp.c:816:17: error: too many arguments to function 'xsk_buff_dma_sync_for_cpu'
       >   816 |                 xsk_buff_dma_sync_for_cpu(xdp, rx_ring->xsk_pool);
       >       |                 ^~~~~~~~~~~~~~~~~~~~~~~~~
       > compilation terminated due to -Wfatal-errors.
       > make[3]: *** [/nix/store/4pi3spm0m1w4hyr659wkpxmf9cj7b6zr-linux-6.10.2-dev/lib/modules/6.10.2/source/scripts/Makefile.build:244: /build/source/kernel/linux/ena/ena_xdp.o] Error 1
       > make[2]: *** [/nix/store/4pi3spm0m1w4hyr659wkpxmf9cj7b6zr-linux-6.10.2-dev/lib/modules/6.10.2/source/Makefile:1934: /build/source/kernel/linux/ena] Error 2
       > make[1]: *** [/nix/store/4pi3spm0m1w4hyr659wkpxmf9cj7b6zr-linux-6.10.2-dev/lib/modules/6.10.2/source/Makefile:240: __sub-make] Error 2
       > make[1]: Leaving directory '/nix/store/4pi3spm0m1w4hyr659wkpxmf9cj7b6zr-linux-6.10.2-dev/lib/modules/6.10.2/build'
       > make: *** [Makefile:102: ena.ko] Error 2
       For full logs, run 'nix log /nix/store/j23sjz0wq8gq3y9p8a6082r0nvg9qnyi-ena-2.12.3-6.10.2.drv'.
       ```

**Comment by Mic92** [comment]
> Okay fixing this patch surfaces more errors. How do we feel about marking this broken on 6.10?

Sounds reasonable. Upstream will catch up eventually.

**Comment by arianvp** [comment]
Okay it builds now on all kernels!

Result of `nixpkgs-review pr 313941` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>5 packages marked as broken and skipped:</summary>
  <ul>
    <li>linuxKernel.packages.linux_lqx.ena</li>
    <li>linuxKernel.packages.linux_xanmod.ena</li>
    <li>linuxKernel.packages.linux_xanmod_latest.ena</li>
    <li>linuxKernel.packages.linux_xanmod_stable.ena</li>
    <li>linuxKernel.packages.linux_zen.ena</li>
  </ul>
</details>
<details>
  <summary>19 packages built:</summary>
  <ul>
    <li>linuxKernel.packages.linux_4_19.ena</li>
    <li>linuxKernel.packages.linux_4_19_hardened.ena</li>
    <li>linuxKernel.packages.linux_5_10.ena</li>
    <li>linuxKernel.packages.linux_5_10_hardened.ena</li>
    <li>linuxKernel.packages.linux_5_15.ena</li>
    <li>linuxKernel.packages.linux_5_15_hardened.ena</li>
    <li>linuxKernel.packages.linux_5_4.ena</li>
    <li>linuxKernel.packages.linux_5_4_hardened.ena</li>
    <li>linuxKernel.packages.linux_6_1.ena</li>
    <li>linuxKernel.packages.linux_6_10.ena</li>
    <li>linuxKernel.packages.linux_6_1_hardened.ena</li>
    <li>linuxKernel.packages.linux_6_6.ena</li>
    <li>linuxKernel.packages.linux_hardened.ena (linuxKernel.packages.linux_6_6_hardened.ena)</li>
    <li>linuxKernel.packages.linux_6_8.ena</li>
    <li>linuxKernel.packages.linux_6_8_hardened.ena</li>
    <li>linuxKernel.packages.linux_6_9.ena</li>
    <li>linuxKernel.packages.linux_6_9_hardened.ena</li>
    <li>linuxKernel.packages.linux_latest_libre.ena</li>
    <li>linuxKernel.packages.linux_libre.ena</li>
  </ul>
</details>

**Pull request merged**